patches/server/1057-add-disguise-api.patch

```diff
@@ -0,0 +1,360 @@
++                            net.minecraft.world.level.GameType.DEFAULT_MODE,
++                            entity.getCustomName(),
++                            null
++                        );
++                    player.connection.send(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(Action.ADD_PLAYER), playerUpdate));
++                    player.connection.send(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(Action.UPDATE_LISTED), playerUpdate));
++                    player.connection.send(create(clientboundAddEntityPacket, net.minecraft.world.entity.EntityType.PLAYER));
++                    if(skinParts != null) {
++                        player.connection.send(new net.minecraft.network.protocol.game.ClientboundSetEntityDataPacket(
++                            clientboundAddEntityPacket.getId(),
++                            List.of(new SynchedEntityData.DataItem<>(Player.DATA_PLAYER_MODE_CUSTOMISATION, (byte) skinParts.getRaw()).value())
++                        ));
++                    }
++                    yield  true;
++                }
++            };
++    }
++
++    /*
++     * Only player disguise needs to be handled specially
++     * because the client doesn't forget the player profile otherwise.
++     * This would result in player being kicked cause the entities type mismatches the previously disguised one.
++     */
++    public static void tryDespawn(ServerPlayer player, Entity entity) {
++        if(entity.getBukkitEntity().getDisguiseData() instanceof PlayerDisguise) {
++            player.connection.send(new ClientboundPlayerInfoRemovePacket(List.of(entity.getUUID())));
++        }
++    }
++
++    private static ClientboundAddEntityPacket create(ClientboundAddEntityPacket packet, EntityType<?> entityType) {
++        return new net.minecraft.network.protocol.game.ClientboundAddEntityPacket(
++            packet.getId(),
++            packet.getUUID(),
++            packet.getX(),
++            packet.getY(),
++            packet.getZ(),
++            packet.getXRot(),
++            packet.getYRot(),
++            entityType,
++            0,
++            Vec3.ZERO.add(packet.getX(), packet.getY(), packet.getZ()).scale(1/8000.0D),
++            packet.getYHeadRot()
++        );
++    }
++
++
++    /*
++     * Is used to skip entity meta that doesn't fit the disguised type.
++     * e.g. Player having a float at index 15 (additional hearts) and the server side entity is an Armorstand
++     * that has a byte at that index.
++     */
++    public static boolean shouldSkip(Entity entity, EntityDataAccessor<?> dataAccessor) {
++        return switch (entity.getBukkitEntity().getDisguiseData()) {
++            case DisguiseData.OriginalDisguise original -> false;
++            case EntityTypeDisguise entityTypeDisguise -> !io.papermc.paper.entity.meta.EntityMetaWatcher.isValidForClass(
++                CraftEntityType.bukkitToMinecraft(entityTypeDisguise.entityType()).getBaseClass(),
```

Comment:
This will always return Entity.class, so the generated content is almost useless and all the metadata are skipped.
Generally i don't think the new module is really needed and this can be done with regular reflection like https://pastes.dev/FNvwYGnwgS

Reply:
I'm against using reflection for that. It can hide potential bugs occuring during runtime. When generating everything during update time we can find issues way faster.

I addressed your concern regarding the base class issue by generating a mapping.

https://github.com/PaperMC/Paper/pull/10478/files#diff-a2c5f4e52ea4c2f48ddb4719a5c436f964a8e49beb30c3c90f099d4b8f94e18a

https://github.com/PaperMC/Paper/pull/10478/files#diff-e150aa8e1c1a506433452cd2d30284e9d7bba7a54f13d8c0faef70e715d816a4R150